File: Samples/CslaNet/cs/DeepData/DeepData.Library/OrderList.cs
using System;
using System.Collections.Specialized;
using System.Configuration;
using System.Data;
using System.Data.SqlClient;
using Csla;
using Csla.Data;

namespace DeepData.Library
{
	public partial class OrderList
	  : ReadOnlyListBase<OrderList, OrderInfo>
	{
		#region " Business Methods "

		///<summary>
		///</summary>
		///<param name="id"></param>
		///<returns></returns>
		public OrderInfo FindById(int id)
		{
			foreach (var item in this)
			{
				if (item.Id == id) return item;
			}
			return null;
		}

		#endregion

		#region " Factory Methods "

		///<summary>
		///</summary>
		///<returns></returns>
		public static OrderList GetList()
		{
			return DataPortal.Fetch<OrderList>();
		}

		private OrderList() { }

		#endregion

		#region " Data Access "

		private void DataPortal_Fetch()
		{
			string dalTypeName = ConfigurationManager.AppSettings["OrderData"];
			if (String.IsNullOrEmpty(dalTypeName))
			{
				Fetch();
			}
			else if (dalTypeName.IndexOf("DAL.Direct") > 0)
			{
				FetchDal();
			}
			else
			{
				FetchDto();
			}
		}

		private void Fetch()
		{
			try
			{
				RaiseListChangedEvents = false;
				IsReadOnly = false;
				using (var cn = new SqlConnection(ConfigurationManager.ConnectionStrings["DeepDataDAL"].ConnectionString))
				{
					cn.Open();
					using (var cm = cn.CreateCommand())
					{
						cm.CommandText = "SELECT id,customer FROM [Order];SELECT orderid,id,product FROM OrderLine;SELECT orderid,lineid,id,detail FROM OrderLineDetail";
						cm.CommandType = CommandType.Text;
						using (var data = new SafeDataReader(cm.ExecuteReader()))
						{
							LoadOrders(data);
						}
					}
				}
			}
			catch
			{
				Exception ex;
			}
			finally
			{
				IsReadOnly = true;
				RaiseListChangedEvents = true;
			}

		}

		private void FetchDal()
		{
			try
			{
				RaiseListChangedEvents = false;
				IsReadOnly = false;
				DeepData.DAL.DataFactory df = new DAL.DataFactory();
				using (var dal = df.GetOrderDataObject())
				{
					SafeDataReader data = ((SafeDataReader)(dal.GetOrders()));
					LoadOrders(data);
				}

			}
			finally
			{
				IsReadOnly = true;
				RaiseListChangedEvents = true;
			}

		}

		private void LoadOrders(SafeDataReader data)
		{
			while (data.Read())
				Add(OrderInfo.GetOrderInfo(data));

			data.NextResult();
			while (data.Read())
				FindById(data.GetInt32("OrderId")).LoadItem(data);

			data.NextResult();
			while (data.Read())
				FindById(data.GetInt32("OrderId")).LoadDetail(data);

		}

		#endregion

	}

}

File: Samples/CslaNet/vb/DeepData/DeepData.Library/OrderList.vb
Imports System.Configuration.ConfigurationManager
Imports System.ComponentModel
Imports System.Data.SqlClient

<Serializable()> _
Public Class OrderList
  Inherits ReadOnlyListBase(Of OrderList, OrderInfo)

#Region " Business Methods "

  Public Function FindById(ByVal id As Integer) As OrderInfo

    For Each item As OrderInfo In Me
      If item.Id = id Then
        Return item
      End If
    Next
    Return Nothing

  End Function

#End Region

#Region " Factory Methods "

  Public Shared Function GetList() As OrderList

    Return DataPortal.Fetch(Of OrderList)()

  End Function

  Private Sub New()

  End Sub

#End Region

#Region " Data Access "

  Private Overloads Sub DataPortal_Fetch()

    Dim dalTypeName As String = AppSettings("OrderData")
    If Len(dalTypeName) = 0 Then
      Fetch()

    ElseIf dalTypeName.IndexOf("DAL.Direct") > 0 Then
      FetchDal()

    Else
      FetchDto()
    End If

  End Sub

  Private Sub Fetch()

    Try
      Me.RaiseListChangedEvents = False
      IsReadOnly = False
      Using cn As New SqlConnection(ConnectionStrings("DeepData.My.MySettings.DeepDataStoreConnectionString").ConnectionString)
        cn.Open()
        Using cm As SqlCommand = cn.CreateCommand
          cm.CommandText = "SELECT id,customer FROM [Order];SELECT orderid,id,product FROM OrderLine;SELECT orderid,lineid,id,detail FROM OrderLineDetail"
          cm.CommandType = CommandType.Text
          Using data As New SafeDataReader(cm.ExecuteReader)
            LoadOrders(data)
          End Using
        End Using
      End Using

    Finally
      IsReadOnly = True
      Me.RaiseListChangedEvents = True
    End Try

  End Sub

  Private Sub FetchDal()

    Try
      Me.RaiseListChangedEvents = False
      IsReadOnly = False
      Dim df As New DeepData.DAL.DataFactory
      Using dal As DeepData.DAL.OrderData = df.GetOrderDataObject
        Dim data As SafeDataReader = CType(dal.GetOrders, SafeDataReader)
        LoadOrders(data)
      End Using

    Finally
      IsReadOnly = True
      Me.RaiseListChangedEvents = True
    End Try

  End Sub

  Private Sub LoadOrders(ByVal data As SafeDataReader)

    While data.Read
      Add(OrderInfo.GetOrderInfo(data))
    End While

    data.NextResult()
    While data.Read
      FindById(data.GetInt32("OrderId")).LoadItem(data)
    End While

    data.NextResult()
    While data.Read
      FindById(data.GetInt32("OrderId")).LoadDetail(data)
    End While

  End Sub

#End Region

End Class

